src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -1363,13 +1363,27 @@ class RsFnInferenceContext(
     }
 
     private fun inferIndexExprType(expr: RsIndexExpr): Ty {
+        fun isArrayToSlice(prevType: Ty?, type: Ty?): Boolean =
+            prevType is TyArray && type is TySlice
+
         val containerType = expr.containerExpr?.inferType() ?: return TyUnknown
         val indexType = expr.indexExpr?.inferType() ?: return TyUnknown
-        return lookup.coercionSequence(containerType)
-            .mapNotNull { type -> lookup.findIndexOutputType(type, indexType) }
-            .firstOrNull()
-            ?.register()
-            ?: TyUnknown
+
+        var derefCount = -1 // starts with -1 because the fist element of the coercion sequence is the type itself
+        var prevType: Ty? = null
+        for (type in lookup.coercionSequence(containerType)) {
+            val outputType = lookup.findIndexOutputType(type, indexType)
+            if (!isArrayToSlice(prevType, type)) derefCount++
+
+            if (outputType != null) {
+                expr.containerExpr?.let { ctx.addAdjustment(it, Adjustment.Deref(containerType), derefCount) }
```

Comment:
I don't sure, but do we N^2 adjustments instead of N? I.e. this `addAdjustment` call is inside a loop, and athere is a loop inside `addAdjustment`